src/ranges.jl

```diff
@@ -0,0 +1,139 @@
+
+"""
+known_first(::Type{T})
+
+If `first` of an instance of type `T` is known at compile time, return it.
+Otherwise, return `nothing`.
+
+@test isnothing(known_first(typeof(1:4)))
+@test isone(known_first(typeof(Base.OneTo(4))))
+"""
+known_first(x) = known_first(typeof(x))
+known_first(::Type{T}) where {T} = nothing
+known_first(::Type{Base.OneTo{T}}) where {T} = one(T)
+
+"""
+known_last(::Type{T})
+
+If `last` of an instance of type `T` is known at compile time, return it.
+Otherwise, return `nothing`.
+
+@test isnothing(known_last(typeof(1:4)))
+using StaticArrays
+@test known_last(typeof(SOneTo(4))) == 4
+"""
+known_last(x) = known_last(typeof(x))
+known_last(::Type{T}) where {T} = nothing
+
+"""
+known_step(::Type{T})
+
+If `step` of an instance of type `T` is known at compile time, return it.
+Otherwise, return `nothing`.
+
+@test isnothing(known_step(typeof(1:0.2:4)))
+@test isone(known_step(typeof(1:4)))
+"""
+known_step(x) = known_step(typeof(x))
+known_step(::Type{T}) where {T} = nothing
+known_step(::Type{<:AbstractUnitRange{T}}) where {T} = one(T)
+
+_eltype(::Type{T}) where {T} = T
+_eltype(::Type{Val{V}}) where {V} = typeof(V)
+
+struct OptionallyStaticRange{T<:Integer,F,S,L} <: OrdinalRange{T,T}
+  start::F
+  step::S
+  stop::L
+
+  function OptionallyStaticRange(start::F, step::S, stop::L) where {F,S,L}
+      T = promote_type(_eltype(F), _eltype(S), eltype(L))
+      return new{T,F,S,L}(start, step, stop)
+  end
+end
+
+Base.first(r::OptionallyStaticRange{<:Any,Val{F}}) where {F} = F
+Base.first(r::OptionallyStaticRange{<:Any,<:Any}) = getfield(r, :start)
+
+Base.step(r::OptionallyStaticRange{<:Any,<:Any,Val{S}}) where {S} = S
+Base.step(r::OptionallyStaticRange{<:Any,<:Any,<:Any}) = getfield(r, :step)
+
+Base.last(r::OptionallyStaticRange{<:Any,<:Any,<:Any,Val{L}}) where {L} = L
+Base.last(r::OptionallyStaticRange{<:Any,<:Any,<:Any,<:Any}) = getfield(r, :stop)
+
+ArrayInterface.known_first(::OptionallyStaticRange{<:Any,Val{F}}) where {F} = F
+ArrayInterface.known_step(::OptionallyStaticRange{<:Any,<:Any,Val{S}}) where {S} =S
+ArrayInterface.known_last(::OptionallyStaticRange{<:Any,<:Any,<:Any,Val{L}}) where {L} = L
+
+function Base.isempty(r::OptionallyStaticRange)
+    return (first(r) != last(r)) & ((step(r) > zero(step(r))) != (last(r) > first(r)))
+end
+
+@inline function Base.length(r::OptionallyStaticRange{T}) where {T}
+    if isempty(r)
+        return zero(T)
+    else
+        if known_step(r) === oneunit(T)
+            if known_first(r) === oneunit(T)
+                return last(r)
+            else
+                return last(r) - first(r) + step(r)
+            end
+        else
+            return Integer(div((last(r) - first(r)) + step(r), step(r)))
```

Comment:
Maybe `Base.udiv_int` if we can guarantee the checks will pass?

Reply:
The code for calculating the length on `StepRange` is what I'm basing this one and I should probably expand this out to also include [this](https://github.com/JuliaLang/julia/blob/master/base/range.jl#L554) sort of code.